Create DiscretizedCurvesSpace

AliceLeBrigant opened the pull request:
I have implemented two different metrics for curves:
- the L2 metric
- the Square Root Velocity (SRV) metric (see e.g. https://ieeexplore.ieee.org/abstract/document/5601739)

The SRV framework is implemented only for curves in a Euclidean space for starters (I have to figure out how to test the fact that the base manifold is Euclidean), because the formulas are considerably simpler in that setting. 

I needed the ```inner_product_matrix``` method of ```Hypersphere``` so I added it.

AliceLeBrigant commented:
Thanks for all the comments ! 

It's not really easy to determine what should the ```dimension``` variable be in ```ParameterizedManifold```. Rigorously, the value should be infinite since the space of curves or surfaces is infinite-dimensional. If we consider sampled curves or surfaces defined by a number ```n```of sample points, then it should be ```n * base_manifold.dimension```, where ```base_manifold``` is the space where the sample points live (a Euclidean space or a Riemannian manifold such as the sphere). But then, there should probably be another variable that distinguishes between curve and surface. For now, ```dimension``` is rather the dimension of the ```ParameterizedManifold``` seen as a _submanifold_ of the ambient space : ```dimension=1``` for a curve and ```dimension=2``` for a surface. What do you think ?

AliceLeBrigant commented on geomstats/parameterized_manifold.py:
```diff
@@ -0,0 +1,158 @@
+
+class L2Metric(RiemannianMetric):
+    """
+    L2 Riemannian metric on the space of parameterized curves or surfaces.
```
I am not sure what you mean by field. In the end, we should end up with one file with
- a class ```ParameterizedManifold```
- a class ```ParameterizedManifoldSpace```
- a class ```L2Metric``` (and other metric classes)
is that correct ?

ninamiolane replied:
By field, I mean "attribute", in the sense of Object Oriented Programming. "Object-oriented programming (OOP) is a programming paradigm based on the concept of "objects", which may contain data, in the form of fields, often known as attributes; and code, in the form of procedures, often known as methods. "

AliceLeBrigant commented on geomstats/parameterized_manifold.py:
```diff
@@ -0,0 +1,158 @@
+    L2 Riemannian metric on the space of parameterized curves or surfaces.
+    """
+    def __init__(self, base_manifold):
+        self.base_manifold = base_manifold
```
I don't really get what is the meaning of the ```signature``` attribute of ```Riemannian Metric``` ?

ninamiolane commented:
Here is what I understood, let me know if I got it right or wrong?

We have a manifold of manifolds, so we have to implement in fact two classes of manifolds:

- the class `ParameterizedManifold`, already implemented: since `points` is given to the constructor, an object instantiated from this class is *one* curve (made of `points`) or *one* surface (made of `points`). The `dimension` of this manifold is 1 for curve, or 2 for surface, it is the dimension of one parameterized manifold: i.e. the `dimension` parameter is easy in this case.

One object of the above class (e.g. a curve) becomes one point in the following manifold:
- the class, not implemented yet, of all curves / all surfaces, i.e. the manifold of the parameterized manifolds. We could call this class `ParameterizedManifoldsSpace` or something like this. Rigorously, `dimension` of this manifold should be infinite indeed. How about we put `dimension = math.inf` for now?

ninamiolane commented:
I'm also thinking: we called this class "ParameterizedManifold", but we don't really have a parameterization so far.

ninamiolane commented:
Yes I agree. In the meantime, since we don't have a parameterization per se, how about moving the methods of the current class ` ` to the class `EmbeddedManifold`?

Yes! :D

A potential problem though, is that it would change the default shape of vectors: which is (n_points, vec_dim) everywhere by convention now. We would need to check if this doesn't impact the code.

Another potential problem with doing [n_tangent_vecs, n_points, dim] is that it might only works when the points on the manifold are represented as vectors, for example a point on the Euclidean space is represented as a vector. If points are represented as matrices, for example a point on a matrix space is represented as a matrix, it might induce confusion because if the exponential output something of shape (10, 3, 3), we don't know if it's a vector case with n_tangent_vecs = 10, n_points=3 and dim=3 or a matrix case with n_tangent_vec = 10 and a matrix of size 3x3.

But we might be able to go around! Could you open a github issue with this? Do you need it soon? Thanks!

AliceLeBrigant commented:
I have rebased to update my branch with the master, deleted ```parameterized_manifold.py``` and created ```discretized_curves_space.py```. 

I haven't forgotten about the ``ìnner_product_matrix``` method of ```HypersphereMetric``` that you asked me to remove, however without it I can't use the ```inner_product``` method, which I need for the same method of the ```L2_metric``` class when the embedding manifold is the sphere. I could call the ``ìnner_product``` method of the embedding manifold instead (R3 in this case), however this only works for embedded manifolds and not Euclidean spaces for example. Do you see a way around this ?

ninamiolane commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,424 @@
+        assert gs.all(velocity_norm != 0)
+        srv = velocity / gs.sqrt(velocity_norm)
+        index = gs.arange(n_points_in_total - 1)
+        index_select = index[(index + 1) % n_sampling_points != 0]
```
what is this index selection?

AliceLeBrigant replied:
This is again linked to vectorization. When ```curve``` has shape ```[n_curves, n_sampling_points, n_coords]```, it has to be reshaped into ```[n_curves * n_sampling_points, n_coords]``` before computing the velocity using
```
velocity = n_sampling_points * self.embedding_metric.log(
                point=curve[1:, :], base_point=curve[:-1, :])
```

At regular intervals of length ```n_sampling_points```, this computes a log that is meaningless, namely the log between the end point of ```curve[k, :, :]``` and the starting point of the next curve ```curve[k + 1, :, :]```. This index selection gets rid of these vectors.

ninamiolane replied:
I see, thanks for explaining! Could you add a comment to explain (very rapidly) this for future collaborators? Thanks!

ninamiolane commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,424 @@
+
+        return log
+
+    def geodesic(self, initial_point,
```
Same: can this function call the parent's function `geodesic` to avoid copy-pasting code?

ninamiolane commented on geomstats/embedded_manifold.py:
```diff
@@ -11,7 +13,8 @@ class EmbeddedManifold(Manifold):
 
     def __init__(self, dimension, embedding_manifold):
-        assert isinstance(dimension, int) and dimension > 0
+        assert isinstance(dimension, int) or dimension == math.inf
```
Nice!

ninamiolane commented on tests/test_discretized_curves_space.py:
```diff
@@ -0,0 +1,250 @@
+
+import geomstats.backend as gs
+
+from geomstats.hypersphere import Hypersphere
```
Alphabetic order

ninamiolane commented on tests/test_discretized_curves_space.py:
```diff
@@ -0,0 +1,250 @@
+ATOL = 1e-8
+
+
+class TestDiscretizedCurvesSpaceMethods(unittest.TestCase):
```
Nice, thanks for the tests!

ninamiolane commented on tests/test_discretized_curves_space.py:
```diff
@@ -0,0 +1,250 @@
+
+N_SAMPLING_POINTS = 10
+SAMPLING_TIMES = gs.linspace(0, 1, N_SAMPLING_POINTS)
+POINT_A = CURVE_A(SAMPLING_TIMES)
```
Maybe `DISCRETIZED_CURVE_A` and same for the others?

ninamiolane commented on tests/test_discretized_curves_space.py:
```diff
@@ -0,0 +1,250 @@
+    def setUp(self):
+        gs.random.seed(1234)
+        self.space_curves_R3 = DiscretizedCurvesSpace(embedding_manifold=R3)
+        self.space_curves_S2 = DiscretizedCurvesSpace(embedding_manifold=S2)
```
Nice!

ninamiolane replied:
space_curves_sphere_2d

ninamiolane commented on tests/test_discretized_curves_space.py:
```diff
@@ -0,0 +1,250 @@
+        self.point_a = POINT_A
+        self.point_b = POINT_B
+        self.point_c = POINT_C
+        points_ab = self.space_curves_S2.L2_metric.geodesic(
```
Maybe defined:
`s2_l2_metric = self.space_curves_s2.l2_metric` since you're using it everywhere?

ninamiolane commented on tests/test_discretized_curves_space.py:
```diff
@@ -0,0 +1,250 @@
+
+    def setUp(self):
+        gs.random.seed(1234)
+        self.space_curves_R3 = DiscretizedCurvesSpace(embedding_manifold=R3)
```
No uppercase letters in variable names. Maybe
`space_curves_in_euclidean_3d` and `space_curves_in_sphere_2d`?

ninamiolane commented on tests/test_discretized_curves_space.py:
```diff
@@ -0,0 +1,250 @@
+class TestDiscretizedCurvesSpaceMethods(unittest.TestCase):
+    _multiprocess_can_split_ = True
+
+    def setUp(self):
```
Actually, this should be `setUpClass(cls)`, otherwise it is run at the beginning and end of each test:
https://stackoverflow.com/questions/23667610/what-is-the-difference-between-setup-and-setupclass-in-python-unittest/23670844

We should correct it elsewhere in geomstats too :)

ninamiolane commented on tests/test_discretized_curves_space.py:
```diff
@@ -0,0 +1,250 @@
+POINT_B = CURVE_B(SAMPLING_TIMES)
+POINT_C = CURVE_C(SAMPLING_TIMES)
+
+N_POINTS = 5
```
This is `N_DISCRETIZED_CURVES`?

ninamiolane commented on tests/test_discretized_curves_space.py:
```diff
@@ -0,0 +1,250 @@
+    def test_srv_metric_geodesic(self):
+        """
+        Test that the geodesic between two curves in a Euclidean space
+        for the srv metric is the L2 geodesic betweeen the curves srvs.
```
Nice!

AliceLeBrigant commented:
Thanks for the review ! While I'm doing the corrections, here are some answers to your general comments:

- Yes you're right, I can use ```to_ndarray``` and it makes the code more readable. Thanks !

- I have indeed re-used the code of the ```geodesic``` method of ```RiemannianMetric```, because I couldn't call it directly. This is because of the issue I told you about in a previous message : it's not possible to compute the exponential map of several vector fields at several base_points for the ```HypersphereMetric``` (whereas it is possible for a ```EuclideanMetric```). If we managed to fix this, I could simply call the method of the parent class.

- Do you mean change ```point``` to ```curve```, ```n_sampling_points``` to ```n_points```, and ```n_points``` to ```n_curves```, in all the methods of the ```L2Metric``` and ```SRVMetric``` classes ?

ninamiolane commented:
- Cool!

- I see, this makes sense.

- Maybe `n_points` to `n_curves` and `point` to `curve`, but keeping the `n_sampling_points`. What do you think? I'm thinking in terms on how easy it would be for us to read the code when we'll come back to it later.

AliceLeBrigant commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,424 @@
+        """
+        assert tangent_vec_a.shape == tangent_vec_b.shape
+        assert tangent_vec_a.shape == base_point.shape
+        n_coord = base_point.shape[-1]
```
In the case of a Euclidean space yes, but in the case of the Hypersphere the coordinates used to define a point are usually exterior, i.e. ```n_coord = embedding_manifold.dimension + 1```. What would be right is ```n_coord = embedding_manifold.embedding_manifold.dimension```. 

After using ```to_ndarray```, the simplest is probably to write:
```
n_curves = tangent_vec_a.shape[0]
n_points = tangent_vec_a.shape[1]
n_coords = tangent_vec_a.shape[2]
```

AliceLeBrigant commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,424 @@
+        base_point = base_point.squeeze()
+        mult_tangent_vecs = 0
+        if tangent_vec.ndim > base_point.ndim:
+            mult_tangent_vecs = 1
```
Indeed it becomes unnecessary when using ```to_ndarray```. It was useful before to handle the case when only one base point is given but several tangent vectors, which is not covered I think by the log method (one base points but several points).

AliceLeBrigant commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,424 @@
+        self.embedding_metric = embedding_manifold.metric
+        self.L2_metric = L2Metric(embedding_manifold=embedding_manifold)
+
+    def pointwise_inner_product(self, tangent_vec_a, tangent_vec_b,
```
It does when tangent vec has shape ```[n_sampling_points, n_coords]``` but not when it has shape ```[n_curves, n_sampling_points, n_coords]```, which I need for the vectorization of my methods.

ninamiolane replied:
I see!

AliceLeBrigant commented:
Yes, keeping ```n_sampling_points``` makes sense !

ninamiolane commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,437 @@
+
+        n_curves = tangent_vec_a.shape[0]
+        n_sampling_points = tangent_vec_a.shape[1]
+        n_coords = tangent_vec_a.shape[2]
```
You can write:
```
n_curves, n_sampling_points, n_coords = tangent_vec_a.shape
```
directly.

ninamiolane commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,437 @@
+        curve_a = gs.to_ndarray(curve_a, to_ndim=3)
+        curve_b = gs.to_ndarray(curve_b, to_ndim=3)
+
+        n_curves = curve_a.shape[0]
```
Same as above.

ninamiolane commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,437 @@
+        tangent_vec = gs.to_ndarray(tangent_vec, to_ndim=3)
+        base_curve = gs.to_ndarray(base_curve, to_ndim=3)
+
+        n_curves = base_curve.shape[0]
```
Same/

ninamiolane commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,437 @@
+        curve = gs.to_ndarray(curve, to_ndim=3)
+        base_curve = gs.to_ndarray(base_curve, to_ndim=3)
+
+        n_curves = curve.shape[0]
```
Same.

ninamiolane commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,437 @@
+        srv_shape = srv.shape
+        srv = gs.to_ndarray(srv, to_ndim=3)
+        n_curves = srv.shape[0]
+        n_sampling_points = srv.shape[1] + 1
```
why +1?

AliceLeBrigant replied:
To be consistent with the names of the other methods : ```n_sampling_points``` is the number of sampling points of the curve, and so when we deal with its derivative or srv (computed with finite differences) we 'loose' a point. I suggest to rename it ```n_sampling_points_minus_one``` ?

This made me realize I should be multiplying by ```n_sampling_points - 1``` rather than ```n_sampling_points``` everywhere where I compute velocities (the curves are supposed to be parameterized between 0 and 1, at regular intervals of 1/(n-1) if there are n points).

ninamiolane commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,437 @@
+                    )
+        srv_shape = srv.shape
+        srv = gs.to_ndarray(srv, to_ndim=3)
+        n_curves = srv.shape[0]
```
maybe use `srv_shape` since you've defined it above? Or get rid of line 268?

AliceLeBrigant replied:
```srv.shape``` doesn't yield the same result as ```srv_shape``` because of the ```to_ndarray``` in between. I am defining ```srv_shape``` to keep the information of the initial shape of ```srv``` so as to give the output in the same shape (e.g. I want different outputs for input shapes ```[1, n_sampling_points, n_coords]``` and ```[n_sampling_points, n_coords]```.

ninamiolane commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,437 @@
+        n_sampling_points = srv.shape[1] + 1
+        n_coords = srv.shape[2]
+
+        srv = srv.reshape(n_curves * (n_sampling_points - 1), n_coords)
```
why -1?

AliceLeBrigant replied:
Same as before.

ninamiolane commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,437 @@
+            self.pointwise_inner_product(tangent_vec_derivative,
+                                         base_curve_velocity,
+                                         base_curve[:, :-1, :])
+        srv_initial_derivative = gs.transpose(coef_1) * \
```
This line is a bit long. Any chance you could decompose the computation, giving meaningful variable names to the auxiliary blocks, and then compute srv_initial_derivative? Thank you in advance!

ninamiolane commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,437 @@
+        srv_initial_derivative = gs.transpose(coef_1) * \
+            gs.transpose(tangent_vec_derivative, (2, 1, 0)) + \
+            gs.transpose(coef_2) * gs.transpose(base_curve_velocity, (2, 1, 0))
+        srv_initial_derivative = gs.transpose(srv_initial_derivative,
```
There seems to be several transposes: aren't cancelling them out, and could we write it directly without tranposing?

ninamiolane commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,437 @@
+                tangent_vec[:, 1:, :] - tangent_vec[:, :-1, :])
+        base_curve_velocity = n_sampling_points * (
+                base_curve[:, 1:, :] - base_curve[:, :-1, :])
+        base_curve_velocity_norm = self.pointwise_norm(base_curve_velocity,
```
Indentation looks a bit weird here?

ninamiolane commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,437 @@
+                                 'Euclidean space.')
+        curve = gs.to_ndarray(curve, to_ndim=3)
+        base_curve = gs.to_ndarray(base_curve, to_ndim=3)
+        n_curves = curve.shape[0]
```
Same:
```
n_curves, n_sampling_points, n_coords = curve.shape
```

ninamiolane commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,437 @@
+        log_derivative = gs.transpose(coef_1) * \
+            gs.transpose(curve_srv - base_curve_srv, (2, 1, 0)) + \
+            gs.transpose(coef_2) * gs.transpose(base_curve_velocity, (2, 1, 0))
+        log_derivative = gs.transpose(log_derivative, (2, 1, 0))
```
Same: several transposes in a row, can we do it without tranposing? Maybe using `gs.einsum` function?

AliceLeBrigant replied:
Yes you're right, it can be done with the ```einsum``` function ! I think I finally understand how it works.

ninamiolane commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,437 @@
+            gs.transpose(curve_srv - base_curve_srv, (2, 1, 0)) + \
+            gs.transpose(coef_2) * gs.transpose(base_curve_velocity, (2, 1, 0))
+        log_derivative = gs.transpose(log_derivative, (2, 1, 0))
+        log_0 = self.embedding_metric.log(point=curve[:, 0, :],
```
maybe `_initial_sampling_points` instead of `_0` would make the code more readable for later?

AliceLeBrigant replied:
How about ```_initial_starting_points``` to be coherent with the input variable name of ```square_root_velocity_inverse``` ?

ninamiolane commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,437 @@
+        log_0 = gs.transpose(np.tile(log_0, (1, 1, 1)), (1, 0, 2))
+
+        log_cumsum = gs.hstack([gs.zeros((n_curves, 1, n_coords)),
+                                np.cumsum(log_derivative, -2)])
```
Use `gs.cumsum` and define `cumsum` in the numpy backend:
https://github.com/geomstats/geomstats/blob/master/geomstats/backend/numpy.py

ninamiolane commented on geomstats/discretized_curves_space.py:
```diff
@@ -0,0 +1,437 @@
+        log_derivative = gs.transpose(log_derivative, (2, 1, 0))
+        log_0 = self.embedding_metric.log(point=curve[:, 0, :],
+                                          base_point=base_curve[:, 0, :])
+        log_0 = gs.transpose(np.tile(log_0, (1, 1, 1)), (1, 0, 2))
```
why do we need this?

AliceLeBrigant replied:
to go from shape ```[n_curves, n_coords]``` to shape ```[n_curves, 1, n_coords]``` to do the the ```cumsum``` afterwards.

ninamiolane commented on geomstats/hypersphere.py:
```diff
@@ -178,6 +178,13 @@ def __init__(self, dimension):
+        """
+        mat = self.embedding_metric.inner_product_matrix(base_point)
+        return mat
+
```
Can we call this method `embedding_inner_product_matrix`?

AliceLeBrigant commented:
Thanks for the review !

If I change the name of the ```inner_product``` method of Hypersphere to ```embedding_inner_product_matrix``` then when I call the ```inner_product``` method of the class ```L2Metric``` when the embedding manifold is a sphere (which itself calls the ``ìnner_product``` method of ```Hypersphere```) I get the error 'The computation of the inner product matrix is not implemented'.

ninamiolane commented:
Ah, I see! 

Then, let's implement the method `inner_product`, using the inner product of the embedding metric, instead of the `inner_product_matrix`. This way, the `inner_product` is the correct method (by definition of the pullback metric) .

(The inner_product_matrix method was incorrect: the dimensionality of the inner_product_matrix on the sphere was not the same as the dimensionality of inner_product_matrix in the Euclidean space.)

Do you agree?

Thanks!

AliceLeBrigant commented:
Do you mean replacing the ```inner_product_matrix``` method of ```Hypersphere``` by an ```inner_product``` method which calls ```self.embedding_metric.inner_product``` ?

ninamiolane commented:
Yes! Would this work?

AliceLeBrigant commented:
Yes it works !
